File: ResidualProgramGenerator.hs
module ResidualProgramGenerator where

import qualified Data.IntMap as IntMap
import qualified Data.Map as Map
import Control.Monad.State

import Algebra
import SLanguage
import ProcessTree

type Sig = (Name, [Name])
type Sigs = IntMap.IntMap Sig

genResidualProgram :: Tree -> (Program, Exp)

genResidualProgram tree =
  (evalState $ genResidualProgram' tree) (IntMap.empty, [])

genResidualProgram' :: Tree -> State (Sigs, [Rule]) (Program, Exp)

genResidualProgram' tree =
  do resExp <- genResPrExp tree (tree IntMap.! 0)
     (_, rules) <- get
     return (Program rules, resExp)

genResPrExp :: Tree -> Node -> State (Sigs, [Rule]) Exp

genResPrExp tree (b@(Node _ bE _ _ bChIds)) =
  case funcAncestors tree b of
    [] ->
      case bE of
        Var _ -> return $ bE
        Call Ctr cname _ ->
          do es <- genResPrExps tree bChIds
             return $ Call Ctr cname es
        Call FCall name args ->
          genResPrCall tree b name args
        Call GCall name args ->
          genResPrCall tree b name args
        Let _ bs ->
          do e' <- genResPrExp tree (tree IntMap.! (head bChIds))
             es' <- genResPrExps tree (tail bChIds)
             let vnames = map fst bs
                 subst = Map.fromList(vnames `zip` es')
             return $ applySubst subst e'
    (Node aId aE aC _ (aChId : _)) : _ ->
      do (sigs, rules) <- get
         let (name, params) = sigs IntMap.! aId
             args = map Var params
             Just subst = matchAgainst aE bE
         case nodeContr (tree IntMap.! aChId) of
           Nothing ->
             return $ applySubst subst (Call FCall name args)
           Just _ ->
             return $ applySubst subst (Call GCall name args)

genResPrExps :: Tree -> [NodeId] -> State (Sigs, [Rule]) [Exp]

genResPrExps tree nIds =
  mapM (genResPrExp tree) (map (tree IntMap.!) nIds)       

isVarTest :: Tree -> Node -> Bool

isVarTest tree (b@(Node _ _ _ _ (bChId : _))) =
  case nodeContr(tree IntMap.! bChId) of
    Nothing -> False
    Just _ -> True

getFGSig :: Tree -> String -> NodeId -> Name -> [Name] ->
              State (Sigs, [Rule]) Sig

getFGSig tree prefix nId name vs =
  do (sigs, rules) <- get
     case IntMap.lookup nId sigs of
       Nothing ->
         do let name' = prefix ++ tail name ++ show (IntMap.size sigs + 1)
                sig' = (name', vs)
                sigs' = IntMap.insert nId sig' sigs
            put $ (sigs', rules)
            return sig'
       Just sig' ->
         return sig'         

putFGRules :: [Rule] -> State (Sigs, [Rule]) ()

putFGRules newRules =
  do (sigs, rules) <-get
     let rules' = rules ++ newRules
     put (sigs, rules')

getChContr :: Tree -> [NodeId] -> [(Name, [Name])]

getChContr tree nIds =
  let children = map (tree IntMap.!) nIds
  in [ (cname, cparams)  |
         Node _ _ (Just (Contraction _ cname cparams)) _ _ <- children]  

genResPrCall :: Tree -> Node -> Name -> [Exp] ->
                  State (Sigs, [Rule]) Exp

genResPrCall tree (b@(Node bId bE _ _ bChIds)) name args =
  let params = vars bE
  in if isVarTest tree b then
       do (sigs, rules) <- get
          (name', _) <- getFGSig tree "g" bId name params
          bodies <- genResPrExps tree bChIds
          let contrs = getChContr tree bChIds
          let grules =
                [GRule name' cname' cparams' (tail params) body' |
                   ((cname', cparams'), body') <- contrs `zip` bodies]
          putFGRules grules                
          return $ Call GCall name' (map Var params)
     else if isFuncNode tree bId then
       do (sigs, rules) <- get
          (name', params') <- getFGSig tree "f" bId name params
          body' <- genResPrExp tree (tree IntMap.! (head bChIds))
          putFGRules [FRule name' params' body'] 
          return $ Call GCall name' (map Var params)
     else
       genResPrExp tree (tree IntMap.! head bChIds)

File: spsc-lite-scala/src/main/scala/spsc/ResidualProgramGenerator.scala
package spsc
import Algebra._

class ResidualProgramGenerator(val tree: Tree) {
  
  private val sigs = scala.collection.mutable.Map[Node, (String, List[Var])]()
  private val defs = new scala.collection.mutable.ListBuffer[Rule]
  lazy val result = (walk(tree.root), Program(defs.toList))
  
  private def walk(n: Node): Term = if (n.funcAncestor == null) n.expr match {
    case v: Var => v
    case Let(_,bs) =>
      var body = walk(tree.children(n).head)
      applySubst(Map(bs map {case (k, _) => k} zip (tree.children(n).tail map walk):_*),
                 body)
    case Ctr(name, _) => Ctr(name, tree.children(n).map(walk))
    case FCall(name, args) => walkCall(n, name, args)
    case GCall(name, args) => walkCall(n, name, args)
  } else sigs(n.funcAncestor) match {
    case (name, args) => 
      if (tree.children(n.funcAncestor).head.contr == null) 
           applySubst(matchAgainst(n.funcAncestor.expr, n.expr), FCall(name, args))
      else applySubst(matchAgainst(n.funcAncestor.expr, n.expr), GCall(name, args))
  }

  def walkCall(n: Node, name: String, args: List[Term]) = {
    val vs = vars(n.expr)
    if (tree.children(n).head.contr != null) {
      val (gname, _) = sigs.getOrElseUpdate(n, (rename(name, "g"), vs))
      for (cn <- tree.children(n)) 
        defs += GRule(gname, cn.contr.pat, vs.tail, walk(cn))
      GCall(gname, vs)
    } else if (tree.leaves.exists(_.funcAncestor == n)) {
      val (fname, fargs) = sigs.getOrElseUpdate(n, (rename(name, "f"), vs))
      defs += FRule(fname, fargs, walk(tree.children(n).head))
      FCall(fname, vs)
    } else walk(tree.children(n).head)
  }
  
  def rename(f: String, b: String) = {b + f.drop(1) + (sigs.size + 1)}
}